File: SundeeFundee/Models/ExerciseDefinitions.swift
import Foundation

enum LiftCategory: String, CaseIterable, Identifiable {
    case squats = "Squat Variations"
    case olympic = "Olympic Lifts"
    case bench = "Bench Press Variations"
    case deadlifts = "Deadlifts"
    case overheadPress = "Overhead Pressing"

    var id: String { rawValue }

    var icon: String {
        switch self {
        case .squats: return "figure.strengthtraining.traditional"
        case .olympic: return "figure.highintensity.intervaltraining"
        case .bench: return "figure.strengthtraining.functional"
        case .deadlifts: return "figure.core.training"
        case .overheadPress: return "figure.arms.open"
        }
    }
}

struct ExerciseDefinition: Identifiable {
    let id: String
    let name: String
    let category: String
    let liftCategory: LiftCategory
    let muscleGroups: [String]
    let description: String
    let supportsStraps: Bool

    init(
        id: String,
        name: String,
        category: String,
        liftCategory: LiftCategory,
        muscleGroups: [String],
        description: String,
        supportsStraps: Bool = false
    ) {
        self.id = id
        self.name = name
        self.category = category
        self.liftCategory = liftCategory
        self.muscleGroups = muscleGroups
        self.description = description
        self.supportsStraps = supportsStraps
    }
}

enum Exercises {

    // MARK: - Squat Variations

    static let squats: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "back-squat-high-bar",
            name: "Back Squat (High Bar)",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "glutes", "core"],
            description: "Bar rests on the traps; keeps the torso more upright, heavily targeting the quads"
        ),
        ExerciseDefinition(
            id: "back-squat-low-bar",
            name: "Back Squat (Low Bar)",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["glutes", "hamstrings", "quadriceps"],
            description: "Bar rests across the rear delts; recruits the posterior chain to move more weight"
        ),
        ExerciseDefinition(
            id: "front-squat",
            name: "Front Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "core", "upper-back"],
            description: "Bar rests on the front rack; requires massive core stabilization and isolates the quads"
        ),
        ExerciseDefinition(
            id: "overhead-squat",
            name: "Overhead Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "shoulders", "core"],
            description: "Bar locked out overhead with a wide grip while squatting; tests shoulder mobility and balance"
        ),
        ExerciseDefinition(
            id: "box-squat",
            name: "Box Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "glutes", "hamstrings"],
            description: "Squat to a box before exploding up; builds starting strength out of the hole"
        ),
        ExerciseDefinition(
            id: "zercher-squat",
            name: "Zercher Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "glutes", "core", "biceps"],
            description: "Bar held in the crook of the elbows; punishes the core and upper back"
        ),
        ExerciseDefinition(
            id: "bulgarian-split-squat",
            name: "Bulgarian Split Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "glutes", "core"],
            description: "Unilateral squat with the rear foot elevated on a bench"
        ),
        ExerciseDefinition(
            id: "goblet-squat",
            name: "Goblet Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "glutes", "core"],
            description: "Dumbbell or kettlebell held at the chest; great for grooving movement patterns"
        ),
        ExerciseDefinition(
            id: "pause-squat",
            name: "Pause Squat",
            category: "squats",
            liftCategory: .squats,
            muscleGroups: ["quadriceps", "glutes", "core"],
            description: "Holding the bottom position for 2–3 seconds to eliminate the stretch reflex"
        ),
    ]

    // MARK: - Olympic Lifts — Clean Family

    static let cleans: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "squat-clean",
            name: "Squat Clean",
            category: "olympic-clean",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "glutes", "traps", "upper-back"],
            description: "Full movement: pulling the bar from the floor and catching it in a deep front squat"
        ),
        ExerciseDefinition(
            id: "power-clean",
            name: "Power Clean",
            category: "olympic-clean",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "glutes", "traps", "upper-back"],
            description: "Catching the bar high (above parallel) rather than riding into a squat"
        ),
        ExerciseDefinition(
            id: "hang-clean",
            name: "Hang Clean",
            category: "olympic-clean",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "glutes", "traps"],
            description: "Starting from the knees or mid-thigh; focuses on the explosive second pull"
        ),
        ExerciseDefinition(
            id: "block-clean",
            name: "Block Clean",
            category: "olympic-clean",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "glutes", "traps"],
            description: "Pulling from elevated blocks to focus on specific phases of the pull"
        ),
        ExerciseDefinition(
            id: "muscle-clean",
            name: "Muscle Clean",
            category: "olympic-clean",
            liftCategory: .olympic,
            muscleGroups: ["traps", "shoulders", "biceps"],
            description: "Pulling and turning the bar over strictly without re-bending the knees"
        ),
    ]

    // MARK: - Olympic Lifts — Jerk Family

    static let jerks: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "split-jerk",
            name: "Split Jerk",
            category: "olympic-jerk",
            liftCategory: .olympic,
            muscleGroups: ["shoulders", "triceps", "quadriceps", "core"],
            description: "Driving the bar overhead while splitting the legs forward and backward"
        ),
        ExerciseDefinition(
            id: "push-jerk",
            name: "Push Jerk",
            category: "olympic-jerk",
            liftCategory: .olympic,
            muscleGroups: ["shoulders", "triceps", "quadriceps"],
            description: "Driving overhead and catching with a slight knee bend, feet parallel"
        ),
        ExerciseDefinition(
            id: "squat-jerk",
            name: "Squat Jerk",
            category: "olympic-jerk",
            liftCategory: .olympic,
            muscleGroups: ["shoulders", "triceps", "quadriceps", "core"],
            description: "Dropping into a full overhead squat to catch the bar; requires elite mobility"
        ),
    ]

    // MARK: - Olympic Lifts — Snatch Family

    static let snatches: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "squat-snatch",
            name: "Squat Snatch",
            category: "olympic-snatch",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "shoulders", "traps", "core"],
            description: "Pulling the bar from the floor and catching it overhead in a full deep squat"
        ),
        ExerciseDefinition(
            id: "power-snatch",
            name: "Power Snatch",
            category: "olympic-snatch",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "shoulders", "traps"],
            description: "Catching the bar overhead without dipping below parallel"
        ),
        ExerciseDefinition(
            id: "hang-snatch",
            name: "Hang Snatch",
            category: "olympic-snatch",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "shoulders", "traps"],
            description: "Starting from the hang position to train the second pull and turnover"
        ),
        ExerciseDefinition(
            id: "block-snatch",
            name: "Block Snatch",
            category: "olympic-snatch",
            liftCategory: .olympic,
            muscleGroups: ["quadriceps", "shoulders", "traps"],
            description: "Starting from blocks to train the second pull and turnover"
        ),
    ]

    // MARK: - Bench Press Variations

    static let benchPress: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "flat-bench-press",
            name: "Flat Barbell Bench Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["chest", "triceps", "shoulders"],
            description: "The standard chest builder"
        ),
        ExerciseDefinition(
            id: "incline-bench-press",
            name: "Incline Bench Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["upper-chest", "shoulders", "triceps"],
            description: "30-45 degree angle; shifts focus to the upper pectorals and anterior deltoids"
        ),
        ExerciseDefinition(
            id: "decline-bench-press",
            name: "Decline Bench Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["lower-chest", "triceps"],
            description: "Bench angled downward; targets the lower chest with reduced range of motion"
        ),
        ExerciseDefinition(
            id: "close-grip-bench-press",
            name: "Close-Grip Bench Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["triceps", "chest", "shoulders"],
            description: "Hands at shoulder width or narrower; heavily shifts load onto the triceps"
        ),
        ExerciseDefinition(
            id: "wide-grip-bench-press",
            name: "Wide-Grip Bench Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["chest", "shoulders"],
            description: "Hands outside standard grip; increases chest stretch"
        ),
        ExerciseDefinition(
            id: "floor-press",
            name: "Floor Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["triceps", "chest"],
            description: "Pressing from the floor; stops elbows at 90° to build triceps lockout strength"
        ),
        ExerciseDefinition(
            id: "spoto-press",
            name: "Spoto Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["chest", "triceps"],
            description: "Pausing the bar 1-2 inches above the chest before pressing to build tension"
        ),
        ExerciseDefinition(
            id: "board-press",
            name: "Board Press",
            category: "bench",
            liftCategory: .bench,
            muscleGroups: ["triceps", "chest"],
            description: "Lowering to a board on the chest to target lockout sticking points"
        ),
    ]

    // MARK: - Deadlifts

    static let deadlifts: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "conventional-deadlift",
            name: "Conventional Deadlift",
            category: "deadlifts",
            liftCategory: .deadlifts,
            muscleGroups: ["hamstrings", "glutes", "lower-back", "traps"],
            description: "Standard stance, hands outside knees; balances load across the posterior chain",
            supportsStraps: true
        ),
        ExerciseDefinition(
            id: "sumo-deadlift",
            name: "Sumo Deadlift",
            category: "deadlifts",
            liftCategory: .deadlifts,
            muscleGroups: ["quadriceps", "glutes", "hamstrings"],
            description: "Ultra-wide stance, hands inside knees; shortens ROM and shifts load to hips and quads",
            supportsStraps: true
        ),
        ExerciseDefinition(
            id: "romanian-deadlift",
            name: "Romanian Deadlift (RDL)",
            category: "deadlifts",
            liftCategory: .deadlifts,
            muscleGroups: ["hamstrings", "glutes"],
            description: "Starting from the top, hinging at the hips with slightly bent knees",
            supportsStraps: true
        ),
        ExerciseDefinition(
            id: "stiff-legged-deadlift",
            name: "Stiff-Legged Deadlift",
            category: "deadlifts",
            liftCategory: .deadlifts,
            muscleGroups: ["hamstrings", "glutes", "lower-back"],
            description: "Similar to RDL but pulling from the floor with straighter legs",
            supportsStraps: true
        ),
        ExerciseDefinition(
            id: "deficit-deadlift",
            name: "Deficit Deadlift",
            category: "deadlifts",
            liftCategory: .deadlifts,
            muscleGroups: ["hamstrings", "glutes", "quadriceps", "lower-back"],
            description: "Standing on a plate or block to increase ROM and build strength off the floor",
            supportsStraps: true
        ),
        ExerciseDefinition(
            id: "trap-bar-deadlift",
            name: "Trap Bar (Hex Bar) Deadlift",
            category: "deadlifts",
            liftCategory: .deadlifts,
            muscleGroups: ["quadriceps", "glutes", "hamstrings", "traps"],
            description: "Hexagonal bar keeps weight centered; a hybrid between a squat and a hinge",
            supportsStraps: true
        ),
    ]

    // MARK: - Overhead Pressing

    static let overheadPress: [ExerciseDefinition] = [
        ExerciseDefinition(
            id: "strict-press",
            name: "Strict Press (Military Press)",
            category: "overhead-press",
            liftCategory: .overheadPress,
            muscleGroups: ["shoulders", "triceps"],
            description: "Standing vertical press with zero leg drive; pure shoulder and triceps strength"
        ),
        ExerciseDefinition(
            id: "push-press",
            name: "Push Press",
            category: "overhead-press",
            liftCategory: .overheadPress,
            muscleGroups: ["shoulders", "triceps", "quadriceps"],
            description: "Using a slight knee dip and drive to launch the weight up, finishing with the arms"
        ),
        ExerciseDefinition(
            id: "z-press",
            name: "Z-Press",
            category: "overhead-press",
            liftCategory: .overheadPress,
            muscleGroups: ["shoulders", "triceps", "core"],
            description: "Pressing while seated on the floor with legs straight; removes all lower body stabilization"
        ),
    ]

    // MARK: - All & Lookup

    static let all: [ExerciseDefinition] =
        squats + cleans + jerks + snatches + benchPress + deadlifts + overheadPress

    static var byCategory: [LiftCategory: [ExerciseDefinition]] {
        Dictionary(grouping: all, by: { $0.liftCategory })
    }

    static func find(byId id: String) -> ExerciseDefinition? {
        all.first { $0.id == id }
    }
}

File: flutter_app/lib/domain/models/exercise_definitions.dart
class ExerciseDefinition {
  const ExerciseDefinition({
    required this.id,
    required this.name,
    required this.category,
    required this.muscleGroups,
  });

  final String id;
  final String name;
  final String category;
  final List<String> muscleGroups;
}

class Exercises {
  static const List<ExerciseDefinition> all = <ExerciseDefinition>[
    // Squat Variations
    ExerciseDefinition(
        id: 'Back Squat',
        name: 'Back Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'glutes', 'hamstrings', 'core']),
    ExerciseDefinition(
        id: 'Front Squat',
        name: 'Front Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'core']),
    ExerciseDefinition(
        id: 'Overhead Squat',
        name: 'Overhead Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'shoulders', 'core']),
    ExerciseDefinition(
        id: 'Box Squat',
        name: 'Box Squat',
        category: 'Squat Variations',
        muscleGroups: ['glutes', 'hamstrings']),
    ExerciseDefinition(
        id: 'Zercher Squat',
        name: 'Zercher Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'core']),
    ExerciseDefinition(
        id: 'Bulgarian Split Squat',
        name: 'Bulgarian Split Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Goblet Squat',
        name: 'Goblet Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'core']),
    ExerciseDefinition(
        id: 'Pause Squat',
        name: 'Pause Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'glutes']),

    // Olympic Lifts - Cleans
    ExerciseDefinition(
        id: 'Squat Clean',
        name: 'Squat Clean',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Power Clean',
        name: 'Power Clean',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Hang Squat Clean',
        name: 'Hang Squat Clean',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Hang Power Clean',
        name: 'Hang Power Clean',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),

    ExerciseDefinition(
        id: 'Block Power Clean',
        name: 'Block Power Clean',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Muscle Clean',
        name: 'Muscle Clean',
        category: 'Olympic Lifts',
        muscleGroups: ['shoulders', 'back']),
    ExerciseDefinition(
        id: 'Clean Pull',
        name: 'Clean Pull',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),

    // Olympic Lifts - Jerks
    ExerciseDefinition(
        id: 'Split Jerk',
        name: 'Split Jerk',
        category: 'Olympic Lifts',
        muscleGroups: ['shoulders', 'triceps', 'legs']),
    ExerciseDefinition(
        id: 'Push Jerk',
        name: 'Push Jerk',
        category: 'Olympic Lifts',
        muscleGroups: ['shoulders', 'triceps']),
    ExerciseDefinition(
        id: 'Squat Jerk',
        name: 'Squat Jerk',
        category: 'Olympic Lifts',
        muscleGroups: ['shoulders', 'legs']),

    // Olympic Lifts - Snatches
    ExerciseDefinition(
        id: 'Squat Snatch',
        name: 'Squat Snatch',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Power Snatch',
        name: 'Power Snatch',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Hang Squat Snatch',
        name: 'Hang Squat Snatch',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Hang Power Snatch',
        name: 'Hang Power Snatch',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),

    ExerciseDefinition(
        id: 'Block Power Snatch',
        name: 'Block Power Snatch',
        category: 'Olympic Lifts',
        muscleGroups: ['full-body']),
    ExerciseDefinition(
        id: 'Muscle Snatch',
        name: 'Muscle Snatch',
        category: 'Olympic Lifts',
        muscleGroups: ['shoulders', 'back']),

    // Bench Press Variations
    ExerciseDefinition(
        id: 'Flat Barbell Bench Press',
        name: 'Flat Barbell Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['chest', 'triceps', 'shoulders']),
    ExerciseDefinition(
        id: 'Incline Bench Press',
        name: 'Incline Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['upper-chest', 'shoulders']),
    ExerciseDefinition(
        id: 'Decline Bench Press',
        name: 'Decline Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['lower-chest', 'triceps']),
    ExerciseDefinition(
        id: 'Close-Grip Bench Press',
        name: 'Close-Grip Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['triceps', 'chest']),
    ExerciseDefinition(
        id: 'Wide-Grip Bench Press',
        name: 'Wide-Grip Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['chest']),
    ExerciseDefinition(
        id: 'Floor Press',
        name: 'Floor Press',
        category: 'Bench Press Variations',
        muscleGroups: ['triceps', 'chest']),

    ExerciseDefinition(
        id: 'Board Press',
        name: 'Board Press',
        category: 'Bench Press Variations',
        muscleGroups: ['triceps', 'chest']),

    // Deadlift Variations
    ExerciseDefinition(
        id: 'Conventional Deadlift (No Straps)',
        name: 'Conventional Deadlift (No Straps)',
        category: 'Deadlift Variations',
        muscleGroups: ['posterior-chain', 'grip']),
    ExerciseDefinition(
        id: 'Conventional Deadlift (With Straps)',
        name: 'Conventional Deadlift (With Straps)',
        category: 'Deadlift Variations',
        muscleGroups: ['posterior-chain']),
    ExerciseDefinition(
        id: 'Sumo Deadlift (No Straps)',
        name: 'Sumo Deadlift (No Straps)',
        category: 'Deadlift Variations',
        muscleGroups: ['posterior-chain', 'quads', 'grip']),
    ExerciseDefinition(
        id: 'Sumo Deadlift (With Straps)',
        name: 'Sumo Deadlift (With Straps)',
        category: 'Deadlift Variations',
        muscleGroups: ['posterior-chain', 'quads']),


    // Hinge Variations
    ExerciseDefinition(
        id: 'Romanian Deadlift / RDL (No Straps)',
        name: 'Romanian Deadlift / RDL (No Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['hamstrings', 'glutes', 'grip']),
    ExerciseDefinition(
        id: 'Romanian Deadlift / RDL (With Straps)',
        name: 'Romanian Deadlift / RDL (With Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['hamstrings', 'glutes']),
    ExerciseDefinition(
        id: 'Stiff-Legged Deadlift (No Straps)',
        name: 'Stiff-Legged Deadlift (No Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['hamstrings', 'lower-back', 'grip']),
    ExerciseDefinition(
        id: 'Stiff-Legged Deadlift (With Straps)',
        name: 'Stiff-Legged Deadlift (With Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['hamstrings', 'lower-back']),
    ExerciseDefinition(
        id: 'Deficit Deadlift (No Straps)',
        name: 'Deficit Deadlift (No Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['posterior-chain', 'grip']),
    ExerciseDefinition(
        id: 'Deficit Deadlift (With Straps)',
        name: 'Deficit Deadlift (With Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['posterior-chain']),
    ExerciseDefinition(
        id: 'Trap Bar / Hex Bar Deadlift (No Straps)',
        name: 'Trap Bar / Hex Bar Deadlift (No Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['posterior-chain', 'quads', 'grip']),
    ExerciseDefinition(
        id: 'Trap Bar / Hex Bar Deadlift (With Straps)',
        name: 'Trap Bar / Hex Bar Deadlift (With Straps)',
        category: 'Hinge Variations',
        muscleGroups: ['posterior-chain', 'quads']),
    ExerciseDefinition(
        id: 'Rack Pull / Block Pull',
        name: 'Rack Pull / Block Pull',
        category: 'Hinge Variations',
        muscleGroups: ['back', 'glutes', 'grip']),

    // Overhead Pressing
    ExerciseDefinition(
        id: 'Strict Press / Military Press',
        name: 'Strict Press / Military Press',
        category: 'Overhead Pressing',
        muscleGroups: ['shoulders', 'triceps']),
    ExerciseDefinition(
        id: 'Push Press',
        name: 'Push Press',
        category: 'Overhead Pressing',
        muscleGroups: ['shoulders', 'triceps', 'legs']),
    ExerciseDefinition(
        id: 'Z-Press',
        name: 'Z-Press',
        category: 'Overhead Pressing',
        muscleGroups: ['shoulders', 'core']),

    // Back Exercises

    ExerciseDefinition(
        id: 'Pull-up',
        name: 'Pull-up',
        category: 'Back Exercises',
        muscleGroups: ['back', 'biceps']),



    // Accessories & Core
    ExerciseDefinition(
        id: 'Walking Lunges',
        name: 'Walking Lunges',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes']),

    ExerciseDefinition(
        id: 'Weighted Planks',
        name: 'Weighted Planks',
        category: 'Core',
        muscleGroups: ['core']),

    ExerciseDefinition(
        id: 'Box Jumps',
        name: 'Box Jumps',
        category: 'Plyometrics',
        muscleGroups: ['legs', 'explosiveness']),


    ExerciseDefinition(
        id: 'Glute Ham Raises',
        name: 'Glute Ham Raises',
        category: 'Accessories',
        muscleGroups: ['hamstrings', 'glutes', 'lower-back']),


    ExerciseDefinition(
        id: 'Broad Jumps',
        name: 'Broad Jumps',
        category: 'Plyometrics',
        muscleGroups: ['legs', 'explosiveness']),
    ExerciseDefinition(
        id: 'Barbell Shrug',
        name: 'Barbell Shrug',
        category: 'Accessories',
        muscleGroups: ['traps']),


    // New Bench Press Accessories
    ExerciseDefinition(
        id: 'Incline Dumbbell Bench Press',
        name: 'Incline Dumbbell Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['upper-chest', 'shoulders']),
    ExerciseDefinition(
        id: 'Dumbbell Bench Press',
        name: 'Dumbbell Bench Press',
        category: 'Bench Press Variations',
        muscleGroups: ['chest', 'shoulders', 'triceps']),
    ExerciseDefinition(
        id: 'Chest Flys',
        name: 'Chest Flys',
        category: 'Accessories',
        muscleGroups: ['chest']),

    ExerciseDefinition(
        id: 'Skullcrushers',
        name: 'Skullcrushers',
        category: 'Accessories',
        muscleGroups: ['triceps']),
    ExerciseDefinition(
        id: 'Barbell Rows',
        name: 'Barbell Rows',
        category: 'Accessories',
        muscleGroups: ['back', 'biceps']),

    ExerciseDefinition(
        id: 'Dumbbell Pullovers',
        name: 'Dumbbell Pullovers',
        category: 'Accessories',
        muscleGroups: ['chest', 'lats']),
    ExerciseDefinition(
        id: 'Lateral Raises',
        name: 'Lateral Raises',
        category: 'Accessories',
        muscleGroups: ['shoulders']),


    ExerciseDefinition(
        id: 'Overhead Dumbbell Triceps Extensions',
        name: 'Overhead Dumbbell Triceps Extensions',
        category: 'Accessories',
        muscleGroups: ['triceps']),
    ExerciseDefinition(
        id: 'Pull-ups',
        name: 'Pull-ups',
        category: 'Accessories',
        muscleGroups: ['back', 'biceps']),
    ExerciseDefinition(
        id: 'Triceps Extensions',
        name: 'Triceps Extensions',
        category: 'Accessories',
        muscleGroups: ['triceps']),

    // Squat Cycle 2 Accessories
    ExerciseDefinition(
        id: 'Good Mornings',
        name: 'Good Mornings',
        category: 'Hinge Variations',
        muscleGroups: ['hamstrings', 'lower-back']),
    ExerciseDefinition(
        id: 'Front-Rack Reverse Lunges',
        name: 'Front-Rack Reverse Lunges',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes', 'core']),

    ExerciseDefinition(
        id: 'Snatch Balance',
        name: 'Snatch Balance',
        category: 'Olympic Lifts',
        muscleGroups: ['shoulders', 'legs', 'core']),

    ExerciseDefinition(
        id: 'Barbell Hip Thrusts',
        name: 'Barbell Hip Thrusts',
        category: 'Hinge Variations',
        muscleGroups: ['glutes', 'hamstrings']),
    ExerciseDefinition(
        id: 'Farmer\'s Carries',
        name: "Farmer's Carries",
        category: 'Accessories',
        muscleGroups: ['grip', 'core', 'traps']),

    // Back Squat Program Accessories
    ExerciseDefinition(
        id: 'Leg Raises',
        name: 'Leg Raises',
        category: 'Core',
        muscleGroups: ['core', 'abs']),
    ExerciseDefinition(
        id: 'Face Pulls',
        name: 'Face Pulls',
        category: 'Accessories',
        muscleGroups: ['shoulders', 'upper-back']),
    ExerciseDefinition(
        id: 'Step-ups',
        name: 'Step-ups',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Russian Twists',
        name: 'Russian Twists',
        category: 'Core',
        muscleGroups: ['core', 'obliques']),
    ExerciseDefinition(
        id: 'Lat Pulldowns',
        name: 'Lat Pulldowns',
        category: 'Back Exercises',
        muscleGroups: ['back', 'lats', 'biceps']),
    ExerciseDefinition(
        id: 'Hollow Body Hold',
        name: 'Hollow Body Hold',
        category: 'Core',
        muscleGroups: ['core', 'abs']),
    ExerciseDefinition(
        id: 'Single-Leg Press',
        name: 'Single-Leg Press',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Knees-to-Chest',
        name: 'Knees-to-Chest',
        category: 'Core',
        muscleGroups: ['core', 'abs']),
    ExerciseDefinition(
        id: 'Heels-Elevated Goblet Squat',
        name: 'Heels-Elevated Goblet Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'core']),
    ExerciseDefinition(
        id: 'Curtsy Lunges',
        name: 'Curtsy Lunges',
        category: 'Accessories',
        muscleGroups: ['glutes', 'quads']),
    ExerciseDefinition(
        id: 'Side Plank',
        name: 'Side Plank',
        category: 'Core',
        muscleGroups: ['core', 'obliques']),
    ExerciseDefinition(
        id: 'Bodyweight Lunges',
        name: 'Bodyweight Lunges',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Bird-Dogs',
        name: 'Bird-Dogs',
        category: 'Core',
        muscleGroups: ['core', 'lower-back']),
    ExerciseDefinition(
        id: 'Air Squats',
        name: 'Air Squats',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Weighted Dead Bug',
        name: 'Weighted Dead Bug',
        category: 'Core',
        muscleGroups: ['core', 'abs']),
    ExerciseDefinition(
        id: 'Leg Press',
        name: 'Leg Press',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Banded Pull-Aparts',
        name: 'Banded Pull-Aparts',
        category: 'Accessories',
        muscleGroups: ['shoulders', 'upper-back']),
    ExerciseDefinition(
        id: 'Hanging Leg Raises',
        name: 'Hanging Leg Raises',
        category: 'Core',
        muscleGroups: ['core', 'abs']),
    ExerciseDefinition(
        id: 'Hack Squat',
        name: 'Hack Squat',
        category: 'Squat Variations',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'Snatch-Grip RDL',
        name: 'Snatch-Grip RDL',
        category: 'Hinge Variations',
        muscleGroups: ['hamstrings', 'glutes', 'upper-back']),
    ExerciseDefinition(
        id: 'Weighted Lunges',
        name: 'Weighted Lunges',
        category: 'Accessories',
        muscleGroups: ['quads', 'glutes']),
    ExerciseDefinition(
        id: 'L-Sit Hold',
        name: 'L-Sit Hold',
        category: 'Core',
        muscleGroups: ['core', 'abs']),
    ExerciseDefinition(
        id: 'Vertical Jumps',
        name: 'Vertical Jumps',
        category: 'Plyometrics',
        muscleGroups: ['legs', 'explosiveness']),
    ExerciseDefinition(
        id: 'Depth Jumps',
        name: 'Depth Jumps',
        category: 'Plyometrics',
        muscleGroups: ['legs', 'explosiveness']),
  ];

  static ExerciseDefinition? findById(String id) {
    for (final ExerciseDefinition exercise in all) {
      if (exercise.id == id) {
        return exercise;
      }
    }

    return null;
  }

  static String nameById(String id) {
    return findById(id)?.name ??
        id
            .split('-')
            .map((word) =>
                word.isNotEmpty ? '${word[0].toUpperCase()}${word.substring(1)}' : '')
            .join(' ');
  }
}
